active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,998 @@
+extension config format. When manually authoring an ARM template, the author must match the expected format as shown in
+the above examples.
+
+Here are some examples of Bicep to ARM template translations:
+
+Bicep:
+```bicep
+extensionConfigs: {
+  k8s: {
+    str: strParam
+    obj: objParam
+    secureStr: secureStrParam
+    secureStrExt: k8s.kubeConfig
+    secureObjExt: k8s.kubeObject
+    ifBothExt: boolParam ? k8s.kubeConfig : k8s2.kubeConfig
+    ifMixed: boolParam ? k8s.kubeConfig : strParam
+  }
+}
+```
+
+ARM template:
+```json5
+{
+  "extensionConfigs": {
+    "k8s": {
+      "str": "[createObject('value', parameters('strParam'))]",
+      "obj": "[createObject('value', parameters('objParam'))]",
+      "secureStr": "[createObject('value, parameters('secureStrParam'))]",
+      "secureStrExt": "[extensionConfigs('k8s').kubeConfig]",
+      "secureObjExt": "[extensionConfigs('k8s').kubeObject]",
+      "ifBothExt": "[if(parameters('boolParam'), extensionConfigs('k8s').kubeConfig, extensionConfigs('k8s2').kubeConfig)]",
+      "ifMixed": "[if(parameters('boolParam'), extensionConfigs('k8s').kubeConfig, createObject('value', parameters('strParam')))]"
+    }
+  }
+}
+```
+
+### Microsoft.Resources/deployments API changes
+
+To support stacks extensibility, the deployments API needs to return additional data about extensible resources.
+It needs to return extensions with their configurations and the resources deployed to those extensions. Extensible resources
+need to be linked to extensions. This can be done with the key: extension name, extension version, and config ID. It
+is possible the key may match to multiple extensions with differing config instructions, but this is an edge case.
+
+Deployment GET:
+```json5
+{
+  // ... deployment contract
+  "extensions": [ // NEW: "Extensions" must be returned for any deployment terminal state (Succeeded, Failed, etc)
+    {
+      "name": "Kubernetes",
+      "version": "1.0.0",
+      "configId": "<ID from extension>", // Provided by the extension in the CREATE/UPDATE return payload.
+      "config": {
+        // the language expressions provided to the PUT have been evaluated
+        "namespace": {
+          "type": "string",
+          "value": "default"
+        },
+        "kubeConfig": {
+          "type": "securestring",
+          // or keyVaultReference
+          "apiReference": {
+            "method": "POST",
+            "armResourceId": "/subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/...",
+            "apiVersion": "2024-02-01",
+            "action": "listClusterAdminCredentials",
+            "query": "a=1&b=2",
+            "responseValuePath": "kubeconfigs[0].value"
+          }
+        }
+      }
+    }
+  ],
+  "outputResources": [ // "outputResources" is only populated when the deployment is successful. Stacks uses operations to peel through failed deployment resources.
+    { // NEW: Extensible resource
```

Comment:
In terms of swagger modeling, if we want to support polymorphism (e.g. "either an ARM resource, or an extension resource"), we may be asked to add a [discriminator property](https://swagger.io/docs/specification/v3_0/data-models/inheritance-and-polymorphism/#discriminator) - e.g. either:
```json
{ "type": "arm", "id": "..." }
```
or:
```json
{ "type": "extension", ...other extension properties... }
```

Reply:
I originally wanted to add this. Right now it's ID vs identifiers, which is a little bit odd. I'm not sure if it's required a discriminator be in that format. Identifiers is also an object, so that complicates things. I'm up for adding it. @shenglol thoughts?

Reply:
I like the idea. Adding a discriminator will improve readability.

Reply:
On second thought, aren't the `extension` properties sufficient enough to distinguish between ARM and extensible resources?

Reply:
Discussed on a call - we should be fine to leave this as-is, and address it if it comes up in review.